src/mlang/m_ir/mir.ml

```diff
@@ -411,7 +411,53 @@ let find_var_by_name (p : program) (name : string Pos.marked) : Variable.t =
            (Pos.VarNameToID.find name p.program_idmap))
     with Not_found -> Errors.raise_spanned_error "unknown variable" (Pos.get_position name))
 
-let find_vars_by_io (p : program) (io_to_find : io) : Variable.t list =
-  VariableMap.fold
-    (fun k v acc -> if v.var_io = io_to_find then k :: acc else acc)
-    p.program_vars []
+let find_vars_by_io (p : program) (io_to_find : io) : rule_data RuleMap.t =
+  let is_var_io (_, var_data) = var_data.var_io = io_to_find in
+  RuleMap.filter_map
+    (fun _ y ->
+      let rule_vars = List.filter is_var_io y.rule_vars in
+      Some { y with rule_vars })
+    p.program_rules
+
+(** Explores the rules to find rule and variable data *)
+let find_var_definition (p : program) (var : Variable.t) : rule_data * variable_data =
+  (* using exceptions to cut short exploration *)
+  let exception Found_rule of rule_data * variable_data in
+  let exception Found_var of variable_data in
+  try
+    RuleMap.iter
+      (fun _ rule_data ->
+        try
+          List.iter
+            (fun (vid, def) -> if var.id = vid then raise (Found_var def))
+            rule_data.rule_vars
+        with Found_var def -> raise (Found_rule (rule_data, def)))
+      p.program_rules;
+    raise Not_found
+  with Found_rule (rule, var) -> (rule, var)
+
+let map_vars (f : Variable.t -> variable_data -> variable_data) (p : program) : program =
+  let program_rules =
+    RuleMap.map
+      (fun rule_data ->
+        let rule_vars =
+          List.map
+            (fun (vid, def) ->
+              let var = VariableDict.find vid p.program_vars in
+              (vid, f var def))
+            rule_data.rule_vars
+        in
+        { rule_data with rule_vars })
+      p.program_rules
+  in
+  { p with program_rules }
+
+let fold_vars (f : Variable.t -> variable_data -> 'a -> 'a) (p : program) (acc : 'a) : 'a =
+  RuleMap.fold
+    (fun _ rule_data acc ->
+      List.fold_left
+        (fun acc (vid, def) ->
+          let var = VariableDict.find vid p.program_vars in
+          f var def acc)
+        acc rule_data.rule_vars)
+    p.program_rules acc
```

Comment:
Nitpicking: rebasing PR branch instead of merging target branch into it avoids bloating the diff.

Reply:
Forget it, github showed me a weird version of the diff form some reason.